### PR TITLE
Add Integration Test for v9 Cluster 

### PR DESCRIPTION
**Pull request by anpanigr**
Here are the usecase
   
   (a) Add a cluster reference (C1)  after domain is started
   (b) Replace a cluster reference (C1) with a new cluster reference (C2)
   (c) Delete a cluster reference (C2)
   (d) Add same cluster reference to two domain to generate "Domain Validation Failed Event"
   (e) Add a (non-existing) cluster reference to a Domain Resource 
   (f) Start a domain with both domain and cluster resource defined in single yaml file

Jenkin Job

https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/13380

https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/13357
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/13358/

https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/13305
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/13306
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/13320/

https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/13310/ ( full run )

### CODE REVIEW

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 1]
```diff
@@ -0,0 +1,606 @@
...
+      checkPodDoesNotExist(managedServerPrefix4 + i,domain4Uid,domainNamespace);
+    }
+
+    // create and deploy cluster resource
+    ClusterResource cluster = createClusterResource(
+        cluster4Res, cluster4Name, domainNamespace, replicaCount);
+    logger.info("Creating cluster {0} in namespace {1}",cluster4Res, domainNamespace);
+    createClusterAndVerify(cluster);
+
+    // check managed server pods are not started
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java:439]:
```suggestion
    // check managed server pods are started
```

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 2]
```diff
@@ -0,0 +1,606 @@
...
+
+    assertTrue(new Command()
+        .withParams(new CommandParams()
+            .command("kubectl create -f " + destDomainYaml))
+        .execute(), "kubectl create failed");
+
+    logger.info("Wait for admin server pod {0} to be ready in namespace {1}",
+        adminPodName, domainNamespace);
+    checkPodReadyAndServiceExists(adminPodName,domain5Uid,domainNamespace);
+    // check managed server pods are not started
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java:486]:
```suggestion
    // check managed server pods are started
```

**Review by vanajamukkara — APPROVED** [review]
LG

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java (hunk 2 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java (hunk 1 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 3]
```diff
@@ -0,0 +1,705 @@
...
+  /*
+   * Create a domain resource DR4 with cluster resource CR4.
+   * Here the cluster resource CR4 refers to WebLogic cluster in model file.
+   * Note here the CR4 is not deployed before deploying resource DR4. 
+   * Deploy resource DR4 and make sure only admin server is started.
+   * (TBD) Should we generate a Warning Event (OWLS-102704) 
+   * Deploy resource CR4 and make sure that all servers in CR4 comes up. 
+   */
+  @Test
+  @DisplayName("Verify servers on missing clsuer resource are picked up when the resource is available")
```
Comment (sankarpn) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java:395]:
```suggestion
  @DisplayName("Verify servers on missing cluster resource are picked up when the resource is available")
```

**Review by sankarpn — CHANGES_REQUESTED** [review]
minor comments

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 4]
```diff
@@ -0,0 +1,705 @@
...
+  private String domain2Uid = "domain2";
+  private String miiImage = null;
+  private static LoggingFacade logger = null;
+  final String adminServerPodName = domainUid + "-admin-server";
+  final String managedServerPrefix = domainUid +  "-c1-managed-server";
+  final String managedServerPrefix2 = domainUid + "-c2-managed-server";
+  final int replicaCount = 2;
+  private String clusterRes = "cluster-1";
+  private String cluster2Res = "cluster-2";
+  private String clusterName = "cluster-1";
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 5]
```diff
@@ -0,0 +1,705 @@
...
+  private static String domainNamespace = null;
+  private String domainUid = "domain1";
+  private String domain2Uid = "domain2";
+  private String miiImage = null;
+  private static LoggingFacade logger = null;
+  final String adminServerPodName = domainUid + "-admin-server";
+  final String managedServerPrefix = domainUid +  "-c1-managed-server";
+  final String managedServerPrefix2 = domainUid + "-c2-managed-server";
+  final int replicaCount = 2;
+  private String clusterRes = "cluster-1";
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 6]
```diff
@@ -0,0 +1,705 @@
...
+   * from cluster-2 comes up.
+   * Delete the cluster Resource (CR2)
+   * Make sure managed servers from cluster-2 goes down 
+   */
+  @Test
+  @DisplayName("Verify dynamic add/remove of cluster resource on domain")
+  void testManageClusterResource() {
+
+    // create and deploy cluster resource(s)
+    ClusterResource cluster = createClusterResource(
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java:176]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 7]
```diff
@@ -0,0 +1,705 @@
...
+import static oracle.weblogic.kubernetes.utils.PodUtils.checkPodExists;
+import static oracle.weblogic.kubernetes.utils.PodUtils.setPodAntiAffinity;
+import static oracle.weblogic.kubernetes.utils.SecretUtils.createSecretWithUsernamePassword;
+import static oracle.weblogic.kubernetes.utils.ThreadSafeLogger.getLogger;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+// Test to create model in image domain with Cluster Resources
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 8]
```diff
@@ -0,0 +1,705 @@
...
+  private static LoggingFacade logger = null;
+  final String adminServerPodName = domainUid + "-admin-server";
+  final String managedServerPrefix = domainUid +  "-c1-managed-server";
+  final String managedServerPrefix2 = domainUid + "-c2-managed-server";
+  final int replicaCount = 2;
+  private String clusterRes = "cluster-1";
+  private String cluster2Res = "cluster-2";
+  private String clusterName = "cluster-1";
+  private String clusterName2 = "cluster-2";
+  private static String configMapName = "cluster-configmap";
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 9]
```diff
@@ -0,0 +1,705 @@
...
+   */
+  @Test
+  @DisplayName("Verify dynamic add/remove of cluster resource on domain")
+  void testManageClusterResource() {
+
+    // create and deploy cluster resource(s)
+    ClusterResource cluster = createClusterResource(
+        clusterRes, clusterName, domainNamespace, replicaCount);
+    logger.info("Creating cluster {0} in namespace {1}",clusterRes, domainNamespace);
+    createClusterAndVerify(cluster);
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java:179]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java (hunk 3 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java:432]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 10]
```diff
@@ -0,0 +1,821 @@
...
+
+    // create encryption secret
+    logger.info("Create encryption secret");
+    createSecretWithUsernamePassword(encryptionSecretName, domainNamespace,
+            "weblogicenc", "weblogicenc");
+  }
+
+  /**
+   * Create a (mii) WebLogic domain DR with domain level replica set to zero.
+   * Do not not associate any Cluster Resource with DR even if two
```
Comment (doxiao) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java:141]: "Do not not" -> "Do not".

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 11]
```diff
@@ -0,0 +1,821 @@
...
+    logger.info("Creating Domain Resource {0} in namespace {1} using image {2}",
+        domainUid, domainNamespace, 
+        MII_BASIC_IMAGE_NAME + ":" + MII_BASIC_IMAGE_TAG);
+    domain.getSpec().withCluster(new V1LocalObjectReference().name(clusterRes));
+
+    createDomainAndVerify(domain, domainNamespace);
+    logger.info("Wait for admin server pod {0} to be ready in namespace {1}",
+        adminServerPodName, domainNamespace);
+    checkPodReadyAndServiceExists(adminServerPodName, domainUid, domainNamespace);
+    // check managed server pods are not started
```
Comment (doxiao) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java:321]: The comment and the code does not match - the comment here says the managed server pods are not started, but the code that followed checks if the managed servers are started.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 12]
```diff
@@ -0,0 +1,821 @@
...
+    for (int i = 1; i <= replicaCount; i++) {
+      checkPodReadyAndServiceExists(managedServerPrefix + i, domainUid, domainNamespace);
+    }
+
+    // Delete the cluster resource with out de-associating it from domain
```
Comment (doxiao) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java:326]: We need to have a test case where the cluster resource is deleted and is de-associated from the domain, and check that the cluster is shut down.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java [hunk 13]
```diff
@@ -0,0 +1,821 @@
...
+        domainNamespace);
+
+    //verify the introspector pod is not started
+    String introspectPodNameBase = getIntrospectJobName(domainUid);
+    checkPodDoesNotExist(introspectPodNameBase, domainUid, domainNamespace);
+
+    String managedServerPrefix = domainUid + "-c1-managed-server";
+    String adminPodName   = domainUid + "-admin-server";
+
+    // check no server pod is not started
```
Comment (doxiao) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java:562]: The comment does not match the code.

**Review by doxiao — COMMENTED** [review]
Looks good over all. Had a couple of comments.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java (hunk 10 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java:141]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java (hunk 13 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java:562]: Modified the comment to "check no server pod is started"

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java (hunk 12 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiClusterResource.java:326]: See the description of the test method testAddReplaceClusterResource()

Here  
  (a) we start the domain with CR1
  (b) replace CR1 with CR2 by patching the domain Resource  ( this is equivalent of de associating cluster Resource)
        and update the restart version 
  (c) Make sure all the server from CR1 gets shutdown and all servers from CR2 comes up

**Pull request merged**